### ISSUE TITLE
New web application now live 🎉

### ISSUE THREAD

**Issue by josdejong** [josdejong/jsoneditor#891]
I've been working on a completely new web application, which is now ~in beta~ live:

https://jsoneditoronline.org/

Please try it out and let me know what you're thinking about it here. Is it easy to use or is it too confusing? Any feedback and suggestions are very welcome!

The most notable changes are:

1. The editor now has two panels each with their own document instead of two panels showing the same document.
2. Local documents are now stored in a local (in-browser) database. When opening the web application in multiple tabs of your browser, each tab will store it's own document separate from the others. No risk of data loss anymore.
3. Support for highlighting differences between two JSON objects.
4. Better support to load a JSON Schema document in various ways: url, text, other panel, document. or embedded text.
5. Better support for drag and drop drag & drop.
6. Fixes in saving large documents to disk (can handle files up to about 500 MB).

**Comment by surajkj** [comment]
One feature I'm missing badly is when we click the second copy i.e. Right to Left, left data in not in formatted mode.

**Comment by aadrian** [comment]
Thank you for the new version. The handling of multiple documents and of huge JSON files is a very nice.

> Any feedback and suggestions are very welcome!

Diffing is very useful, however seems to be restricted to tree mode only.

It would be even more useful if it were active for code mode too. Code editors already have
code diff (and merge) implemented https://codemirror.net/demo/merge.html or http://ace-diff.github.io/ace-diff/
so it would just need integration.

**Comment by josdejong** [comment]
> One feature I'm missing badly is when we click the second copy i.e. Right to Left, left data in not in formatted mode.

Thanks for your feedback @surajkj , that should be easy to fix 👍

**Comment by josdejong** [comment]
> Diffing is very useful, however seems to be restricted to tree mode only.
> 
> It would be even more useful if it were active for code mode too. Code editors already have
> code diff (and merge) implemented https://codemirror.net/demo/merge.html or http://ace-diff.github.io/ace-diff/
> so it would just need integration.

That's a great idea, thanks @aadrian. I've indeed started implementing this for tree mode. I also want to improve the current diffing such that it can recognize items which are inserted or removed from an array.

**Comment by josdejong** [comment]
@surajkj the formatting when copying from left to right panel or vice versa should now be fixed (you probably need to refresh your browser with Ctrl+F5).

**Comment by AjitViolet** [comment]
Nice Upgrade. Best bulk JSON Data Management.

**Comment by josdejong** [comment]
Thanks Ajit :+1:

**Comment by MishaKav** [comment]
Beta is really nice.
I think it would be much better if you would add support for tabs (like in Chrome or any IDE). It much more easy to see all json's data in same site in different tabs instead many opened tabs in Chrome with different data.

**Comment by josdejong** [comment]
Thanks for your feedback and suggestion Misha, I've created a separate issue to discuss in detail whether or not we want to have tabs in #898.

**Comment by josdejong** [comment]
The new web application is now live :smile:, see https://jsoneditoronline.org

**Comment by Itazake** [comment]
In the "about" section, the link to "see changelog" is throwing : 404 not Found NoSuchKey
Ask if need more details, have a nice day

**Comment by josdejong** [comment]
Thanks for reporting @Itazake . Can you see if the issue still occurs after you refresh the web application https://jsoneditoronline.org with Ctrl+F5? The link should open https://jsoneditoronline.org/changelog.html.

**Comment by Itazake** [comment]
Indeed, it work's now, thanks !

**Comment by josdejong** [comment]
Good to hear, thanks for checking.

**Comment by kapuraghav** [comment]
Hi! I really like your app. but now I can not access the recently saved json files, although when I check the localstorage of the browser ( via developer tool) I do see all of the locally saved json files. is there a way to access those in your new app?

**Comment by josdejong** [comment]
Thanks for your feedback @kapuraghav . The list recent files that you've saved in the cloud should be available in the new application when selecting "Open", "Open recent file". In the old application the list was stored in `localStorage.files`, in the new application as `localStorage.recentFiles`.

In the old web application there was only one recent file saved locally in your browser, in `localStorage.data`. This should be openend when opening the new application for the first time.

The legacy list with recent files and opened file are migrated only once when opening the new web application for the first time. After that, a flag `localStorage.firstTimeUse='false'` is set. If you remove this flag from localStorage and load the new web application again it should migrate the list again.

In the worst case you can copy the id's from the files that you have in `localStorage.files` and open them by id in a new browser tab, like: `https://jsoneditoronline.org/#left=cloud.1234` where you have to replace `1234` with the id of your file.

Does this help?

**Comment by IvMisticos** [comment]
I love it, it is great. Thanks

**Comment by josdejong** [comment]
Thanks Iv, good to hear :)

**Comment by sagarhani** [comment]
Good to see some new features. Thanks. 🤩

**Comment by babakarj** [comment]
Thanks for the new design :)
Would you add text size or character count, too?

**Comment by abhilashv1** [comment]
I am really loving this updated design it is very helpful,  appreciate your efforts!

**Comment by josdejong** [comment]
Thanks :)

> Would you add text size or character count, too?

Yes that is a good idea! I was thinking about extending the current poup where you can set the name of the document such that it shows all kind of meta data: when was it last changed and how big the file is.

**Comment by ari-codes** [comment]
I use this almost daily and its awesome. 
I can see ' to " can be solved clicking tool button but what about "True","False",None?
As I use python i need to replace manually :(

**Comment by spirostz** [comment]
Congratulation for the new design and features... they are great!
Just a minor problem to report:
When you are on Firefox private browsing, the document section on top is failing with the below error:
> InvalidStateError: A mutation operation was attempted on a database that did not allow mutations.

**Comment by steveaw1** [comment]
Excellent new features - thank you !
One issue is that - at least in Chrome v79 on Mac - select all (prior to paste new content) doesn't work easily
I'm used to Cmd-A,  delete, - Cmd-V to do this in the old version, but I guess a combined button, as long as this was clear, would be a good substitute.    Clear + paste into either pane would be ideal

**Comment by Carlosferrerhernandez** [comment]
Thank you !

**Comment by josdejong** [comment]
Thanks for all inputs! Really helpful.

----

> I can see ' to " can be solved clicking repair button but what about "True","False",None?
> As I use python i need to replace manually :(

@ari-codes I've opened a separate issue to discuss the details: #906

----
> When you are on Firefox private browsing, the document section on top is failing with the below error:
> > InvalidStateError: A mutation operation was attempted on a database that did not allow mutations.


@spirostz that's an interesting point, makes sense. I will create a better readable error message when working in private mode, say something like "Can't save the document locally (Are you browsing in private mode?)".

----
> One issue is that - at least in Chrome v79 on Mac - select all (prior to paste new content) doesn't work easily

@steveaw1 just to be sure I understand you: The Ctrl+A, Ctrl+V should still work when in `code` mode, and it does not work when in `tree` mode (whilst the latter would be very handy). Do you key combinations work normally in `code` mode?

**Comment by josdejong** [comment]
@spirostz I've deployed a fix which gives a more readable error message when your browsing in private mode on Firefox. Web application version: 5.1.0.

**Comment by Zumbalamambo** [comment]
@josdejong  I have been using this since 2016. The current update kills UX. I didn't know that there was "Code" and "Tree" option on the top. It was seamless. The past app was more easy to work with.

**Comment by kapuraghav** [comment]
@josdejong  Thank you!

**Comment by josdejong** [comment]
> @josdejong I have been using this since 2016. UX could be improvised a bit. I didn't know that there was "Code" and "Tree" option on the top. It was seamless. The past app was more easy to work with.

@Zumbalamambo thanks for your feedback, this is very useful! Can you elaborate a bit on why it's less easy to work with now? Is it because you can toggle mode code/tree now on both sides? Or is it because there are two editors now? Something else?

**Comment by YogeshSomawar** [comment]
Too good... New UI looks awesome & more user-friendly.

**Comment by DheerendraRathor** [comment]
These new changes are amazing! Best part is saving files locally in browser and accessing them from browser.
This has been my go-to tool for all json editing so far!

few feature request though: 
- Folder structure in locally saved files. So it will be easy to organize
- Search option in saved files
- Export data option, so I can move it between browsers and sync to external services like google drive

**Comment by narendravardi** [comment]
@josdejong lot of great features implemented in the new version. Thanks a lot! :)

> Any feedback and suggestions are very welcome!

Talking about this, I some how feel that green colour (#6F9970) is not matching with the orange color (#F4A460). This is my perspective. Please take more inputs and take action accordingly! :)

**Comment by rahul-mutyalkar** [comment]
Sometimes I get this error when I opens shareable link
![json error](https://user-images.githubusercontent.com/11543852/74805914-9999e700-530a-11ea-8cb7-3b27fa107ece.png)

**Comment by josdejong** [comment]
@DheerendraRathor thanks, you have some great suggestions! I really love the idea of linking with an external service like Google Drive or Dropbox or so to store your data. It would be also nice to be able to export all your documents. I'm not sure if we should start creating a folder structure right now, but adding a search box is a very good idea. I've created two separate issues to look into exporting and linking to an external service: #909 and #910, and will look into creating a search box soon.

@narendravardi you may be right about the color mismatch. I will have a look to see if I can tune that. And any suggestion is very welcome too!

@rahul-mutyalkar the error "Local document not found" is probably a local document stored in someone else's browser. That can't be shared. I'm planning on creating a "share" button in each of the two panels, which will only share a document stored in the cloud. I think that will help.

**Comment by josdejong** [comment]
Just published `v5.2.0` with some new, requested features.

@babakarj there is now a new modal "Document Properties" under "Options", which amongst others shows the document size. I still want to merge this with the current "Rename" modal, but one step at a time.

@DheerendraRathor the "Open a recent file" modal now has a search box on top :)

@rahul-mutyalkar I've created a new menu item "Share" which creates a sharable url for you of just one of the two panels instead of both.

**Comment by Vandesm14** [comment]
The new version is awesome! Great job!
The only feature I find missing is to display JSON data as a table.

**Comment by josdejong** [comment]
Thanks @Vandesm14 . Table view would indeed be cool :sunglasses:. There is an open issue for it, see #402.

**Comment by josdejong** [comment]
@narendravardi I've tweaked the color schema by generating a color scheme online: https://coolors.co/7fa360-f4a259-f4e285-5b8e7d-e55b62 (published in `v5.2.2`). Is this better in your opinion? If not, do you have suggestions?

**Comment by meindonnerstag** [comment]
["awesome", ''great", "fantastic"]

**Comment by josdejong** [comment]
:joy: `{"thank": "you"}`

**Comment by PSeitz** [comment]
I'd prefer the old default to have text on the left and the tree on the right.

**Comment by josdejong** [comment]
Thanks for your feedback PSeitz. This is still the default (text left, tee right), except that you can now switch. I understand you like the old, simpler interface where there was no option to switch at all.

**Comment by PSeitz** [comment]
Ah I didn't notice because I pasted at first an invalid json (missing brackets) and it switched to code and stayed that way. I'm not sure if that is intended.

**Comment by josdejong** [comment]
If you load a document which contains invalid json, the editor indeed switches to code-mode (since tree-mode cannot load invalid json). I do understand you confusion though - the switch is done automatically and you may not have a clue how to switch back to tree-mode. I have to give that some more thought.

**Comment by Shredsauce** [comment]
Nice job, I love it even more now!

**Comment by narendravardi** [comment]
@josdejong what do you think about using #265A56 instead of existing green ?
<img width="1680" alt="Screenshot 2020-02-24 at 6 14 19 PM" src="https://user-images.githubusercontent.com/20625060/75153534-a31fb680-5731-11ea-9163-1be43ecb2ee0.png">

**Comment by josdejong** [comment]
Thanks for your suggestion @narendravardi. Now I do have the feeling these two colors you propose do not really fit in the same palette 😂 . Btw I'm not perse bound to the sandybrown menu color though, it could be two totally different colors. I do like warm colors though as opposed to a neutral, "gray mouse" color palette. I really like the palettes that  https://coolors.co comes up with, did you also give that a try?

**Comment by vanduc1102** [comment]
Hi @josdejong ,

Thanks for the update, I TURN-OFF my adblocker on your site! 

i works with Large JSON file daily, i would like to have more specific ERROR,
![image](https://user-images.githubusercontent.com/3014017/75226873-b4270100-57df-11ea-9b04-8889e01e8ef5.png)

I'm was very happy with the previous version at the point.
Could you please check?

**Comment by derskeal** [comment]
The new design is great. The new features are awesome!

**Comment by josdejong** [comment]
@derskeal thanks :+1: 

@vanduc1102 thanks for supporting me :smile: You can only switch to "tree" mode when your JSON is valid (also in the old application, the tree mode could only render valid JSON). In your screenshot I see a red error below line 7, can you try to fix that and try again? (maybe the "repair" button can help you)

**Comment by josdejong** [comment]
@vanduc1102 your suggestion makes sense, I will add the actual parse error to the message.

**Comment by gearintellix** [comment]
best json editor online tools, awesome!

**Comment by philjones** [comment]
@josdejong the new diff mode is really promising!

One thing that would really make it so much better would be the ability to copy a partial item.

Imagine a scenario where there are a bunch of changes that you don't want and a few that you do
![image](https://user-images.githubusercontent.com/692676/75288611-b5a9f680-57ea-11ea-8b32-b01fcfb724da.png)

The overall copy button doesn't work because it takes everything but if you had the ability to copy a node to the other side that would be amazing.

Even if it was just "copy node as JSON" so that you could switch the other tree to code and paste it where you need it.

Additionally, for large documents it can be hard to "find the same spot" when you switch back to code so it would be so cool for the switch to code put you at the same place in the document as you were in "tree".

**Comment by Zumbalamambo** [comment]
@josdejong its because those buttons were too small. especially the white overlay hides the text and makes it seamless... For few moments I couldn't figure out where things are :P

**Comment by josdejong** [comment]
@philjones thanks that are useful suggestions! I would love to have a real cut/copy/paste support, but that's not trivial to implement. Just a "Copy to clipboard" for a certain part of the JSON tree would be easy to implement. I've created a separate issue to address this idea: #914.

I think your idea of getting back to where you where when switching between code/tree makes a lot of sense. It should ideally also remember which nodes you had expanded/collapsed. I have to think about how we can achieve that. See also #899.

@Zumbalamambo I will give it some more thought. The buttons are small but that's not so easy to change because it's part of the JSONEditor library. As soon as you know it it's not an issue anymore I guess. Maybe I can create something to help you discover this, like highlight the menu buttons for a couple of seconds after automatically switching to code mode.

**Comment by josdejong** [comment]
@vanduc1102 the latest version (`v5.2.4`) now shows a clear error message when switching to tree mode fails because of invalid JSON.

**Comment by vanduc1102** [comment]
Great @josdejong 

![image](https://user-images.githubusercontent.com/3014017/75407766-5eb93400-5946-11ea-96a9-d834b2506686.png)

thanks you so much.

**Comment by josdejong** [comment]
👍  😄

**Comment by amreniouinnovent** [comment]
@josdejong I like black gray colors :)
I think we may need dark mode

**Comment by josdejong** [comment]
Thanks for your suggestions @amreniouinnovent . A dark theme would be cool indeed, though it's not the highest prio at this moment. We've an issue open for it: #893

**Comment by QuadCoreBr** [comment]
Hi! 
I really like the new look & feel of your app, I use it almost daily. 
I only have one suggestion for the app, I would like to have the option to colorfy the JSON when pasting it in Google Docs files, I've been using the following app to do such thing http://www.bodurov.com/JsonFormatter/, so I copy and paste the JSON from your app to the other and then I paste the JSON with the colors in my documents. Is it possible to do the same with yours? Avoiding that extra step (call me lazy then haha).

Thank you!

**Comment by josdejong** [comment]
Thanks @QuadCoreBr , that's a nice idea! I've created a separate issue for your suggestion to address it: #924. It's not the highest prio at this moment but I like the idea (also, being able to copy the document somehow when in tree mode would be very helpful).

**Comment by josdejong** [comment]
I've just published an improvement in `v5.3.0` that I'm very excited about: proper array diffing.

Before, arrays where compared simply by index (`isEqual(left[i], right[i])`):

![image](https://user-images.githubusercontent.com/568626/75899757-3c3e8380-5e3c-11ea-9a34-49ea48ac2c0f.png)

Now, we have proper array diffing (powered by [diff-sequences](https://www.npmjs.com/package/diff-sequences)), which then looks like:

![image](https://user-images.githubusercontent.com/568626/75899806-4d879000-5e3c-11ea-9078-ef9d32f71f96.png)

**Comment by spirostz** [comment]
Hi again!
A minor problem with the sorting of elements
Please see image for reproducing:
![image](https://user-images.githubusercontent.com/11974067/75998934-3911c900-5f0a-11ea-824f-a573d0697a8c.png)

**Comment by josdejong** [comment]
Thanks for your feedback and clear explanation @spirostz. The reason here is that when you copy an updated version to the right side, existing fields are not updated and new fields are simply appended to the end of the object. Would be nice indeed if this update also adjusts the order of the keys. I've openend a separate issue to address this: #926.

**Comment by sigalou** [comment]
Hello, I tried to import the official Amazon file for smartHome and the import does not work. The JSON tree cannot be drawn.

The file is alexa_smart_home_message_schema.json and it can be found here:
https://github.com/alexa/alexa-smarthome/blob/master/validation_schemas/alexa_smart_home_message_schema.json

[alexa_smart_home_message_schema.zip](https://github.com/josdejong/jsoneditor/files/4319522/alexa_smart_home_message_schema.zip)

Can you test?

**Comment by josdejong** [comment]
@sigalou thanks for your report. I've loaded the schema and it works just fine as far as I can see, I get proper schema error messages. Can it be that you try to load the [github webpage](https://github.com/alexa/alexa-smarthome/blob/master/validation_schemas/alexa_smart_home_message_schema.json) of the JSON document instead of the [raw JSON document](https://raw.githubusercontent.com/alexa/alexa-smarthome/master/validation_schemas/alexa_smart_home_message_schema.json) itself?

**Comment by josdejong** [comment]
> @josdejong its because those buttons were too small. especially the white overlay hides the text and makes it seamless... For few moments I couldn't figure out where things are :P

@Zumbalamambo I've made a small change in `v5.3.1` such that the mode toggle will blink a few times when the editor automatically switches from tree to code mode. I hope that will help to understand what's going on.

**Comment by josdejong** [comment]
It's time to close this issue now, the new application is live for a month now. If you've more ideas, suggestions, or bugs, please create a new issue for it.

Thanks for all your feedback and support :+1:

**Comment by josdejong** [comment]
@amreniouinnovent it has been a while, but https://jsoneditoronline.org finally has support for a dark theme (v6.1.0). You may need to refresh with Ctrl+F5 to load the newest version in your browser.

![afbeelding](https://user-images.githubusercontent.com/568626/168316021-f4062184-ed27-44ee-bffc-6435cae56198.png)
